### ISSUE TITLE
[FEAT] Add ability to do efficient blocking based on list/array intersections 

### ISSUE THREAD

**Issue by nerskin** [moj-analytical-services/splink#1448]
It’s very common in our context for an entity to have many-valued
attributes. We would like to store these as columns of lists/arrays, and
to be able to block on these list-valued attributes by having blocking
rules that include a pair of records if their list-valued attributes
intersect.

This is currently possible in splink by using the same trick as used in
array_intersect_at_sizes. For example, in duckdb the blocking condition
would look something like

    list_unique(l.list_column) + list_unique(r.list_column) > list_unique(list_concat(l.list_column,r.list_column))

This is extremely slow, however, because the backend will not recognise
that it is possible to perform the join by placing the individual values
in an inverted index. The execution plan for this will just be a cross
join followed by a filter (i.e. O(n^2) complexity), so it currently
isn’t practical to do this for any reasonably large datasets.

### Describe the solution you’d like

My proposal is to allow for this by unnesting/exploding the list-column
before blocking on the variable. The steps for each blocking rule are:

-   Unnest the list-valued columns to create a pair of long tables with
    the id columns and the blocking variables used for the join.
-   Inner join the input tables on the blocking variables
-   Obtain the set of distinct id pairs that appear in the result of
    this join
-   Join the rest of the variables back on to obtain the appropriate
    blocked table

This will allow the query engine to recognise that the query can be done
as a hash join.

I have a rough implementation of this idea for the duckdb backend in a
fork at https://github.com/nerskin/splink (see the README for a simple
example).

If you think this would be a useful feature, I can work on preparing a
pull request. If so, I would appreciate some opinions on how best to
expose this option to the user, and on where the new logic should live -
it might not make sense to keep it in splink/blocking.py, since the
syntax for unnesting/exploding differs between SQL backends.

### Describe alternatives you’ve considered

It could be useful to add an option for the user to provide a table of
pairs of ids that should be compared, instead of or in addition to
providing conventional blocking rules. This would allow users to
implement unconventional blocking rules like this one on their own and
then pass the pairs to splink to do the comparisons.

**Comment by zmbc** [comment]
@nerskin I'm just curious: have you considered exploding the column before passing the data to splink in the first place? For example making 1 row with 10 items in "list_column" into 10 rows with a single item each in "list_column" and identical values in all the other columns?

It seems like this would allow you to block on the column as you want, _and_ it would also allow you to compare similarity on that column as well, which your solution doesn't allow. But maybe it is not feasible with large datasets and long lists to have that much duplication?

**Comment by nerskin** [comment]
@zmbc My solution definitely does allow for comparing similarity on the column (this is done in the demo at https://github.com/nerskin/splink/blob/master/README.md)

unnesting first would cause other problems 

* It makes the input tables unreasonably big, as you suggest, and results in lots of redundant comparisons  
* It interferes with parameter estimation
* It makes it harder to account for the full degree of overlap between the list columns in the later steps (i.e. you can't then use `array_intersect_at_sizes` when calculating the weights), which is what I ultimately want to do.

**Comment by zmbc** [comment]
@nerskin Good points all. It is interesting to see how differently we are looking at this problem, and I think it mostly has to do with the form of our data.

In your case, your array contains IDs, so the only thing interesting about them is whether they are exactly the same. ID 4 isn't "close" to ID 5, nor to ID 44. So, as you say, `array_intersect_at_sizes` is exactly the similarity comparison that you want.

My array contains strings that represent people's names. So I'm not only interested in exact matches, but also similar strings like "Jeff" and "Jeffrey." I'm not aware of a way to do this between _sets_ of names, which is why I said your approach wouldn't allow me to compare on the column, only block on it.

You are definitely right that unnesting first makes the tables big.

I'm not sure I agree that it interferes with parameter estimation. Of course, all parameter estimation should be done with the unnested version of the dataset, so for example the prior probability of a match will be lower. It does violate conditional independence of columns, but so do many real-life situations anyway. I've been able to use EM pretty successfully with unnested datasets.

Again, I'm not aware of a metric that compares string similarity between sets of strings. Your approach with `cl.array_intersect_at_sizes("random_id",[1])` amounts to "does at least one element overlap," and the approach of comparing unnested datasets amounts to "is at least one element similar at the desired similarity level" using e.g. Levenshtein distance.

**Comment by RobinL** [comment]
Hey @nerskin.   

I agree this would be super useful so it'd be great to work with you on this.   Internally, we have faced this kind of problem of many occasions.  

Reply quickly this evening with some initial thoughts, but i'll continue to think about this a bit more deeply.








### 

We've also had similar discussions as above with @zmbc about whether to explode out everything before linking, or write every more complex array comparison formulas.  
Overall, I think there are pros and cons of both approaches.  Neither is perfect and the best solution probably depends on your specific dataset.

But your issue is specifically around blocking, which is quite nice because it kindof sidesteps the above.   There's are definitely some scenarios where it's useful to to block on array columns, and currently this is not possible for performance reasons.

**Comment by zmbc** [comment]
> The alternative that comes to mind is to explode, link, cluster links using connected components, create a new variable (say array_block_var in the input dataset) of the clusters.

Perhaps this is intentional, but this is not equivalent to what @nerskin described. Consider row A with a value of [1, 2] in the array column, row B with a value of [2, 3], and row C with a value of [3, 4]. If you take the links with any intersection and cluster using connected components, row A and row C are in the same cluster despite their values having no intersection.

**Comment by nerskin** [comment]
Hi @RobinL 

Here is the relevant part of the generated SQL from the example in my fork of Splink (slightly simplified). This example is for a deduplication, and the blocking rule is using a list-column called `random_id` and `gender`, a regular column. 
```
(WITH __splink__df_concat_with_tf as (select * from __splink__df_concat_with_tf_574a1f956), 
__splink__df_blocked as (
        (with unnested___splink__df_concat_with_tf as (select *,unnest(random_id) as unnested_random_id from __splink__df_concat_with_tf),
        pairs as (select distinct l.unique_id as left_unique_id,r.unique_id as right_unique_id 
                        from unnested___splink__df_concat_with_tf as l inner join unnested___splink__df_concat_with_tf as r 
                        on l.unnested_random_id = r.unnested_random_id and l.gender = r.gender and  l."unique_id" < r."unique_id"),
        left_df as (select * from 
                        pairs left join __splink__df_concat_with_tf
                        on pairs.left_unique_id = __splink__df_concat_with_tf.unique_id)
select "l"."unique_id" as "unique_id_l", "r"."unique_id" as "unique_id_r", [many other output columns],'0' as match_key  
                from left_df as l left join __splink__df_concat_with_tf as r
                on l.right_unique_id = r.unique_id)
)
```
 
* `unnested__splink__df_concat_with_tf`, is an unnested version of the `__splink__df_concat_with_tf` table used in the internals of splink.
* `pairs` is a table containing the set of unique ids returned by inner-joining the unnested table to itself on the blocking variables. This will contain every distinct pair of `unique_id`s for which their gender columns agree and their list-valued `random_id` columns overlap.
* `left_df` is the result of joining the non-blocking columns from the left dataset back onto `pairs`
* The final result is just the `left_df` with the columns from the right input dataset added back (which is the same as the left dataset in this case since it's a deduplication).

> I believe your approach is effectively to create a 'link table' that creates the pairwise IDs that result from exploding and joining on the exploded data.

This is basically it.

> I don't feel I have a very good intuition for how to write this in a way that SQL engines will be able to process efficiently (especially for larger datasets). So I think we might want to do a bit of experimentation before deciding on an implementation. (Though on the other hand, I can also see an argument for getting something working as a starting point, and then iterating).

> In particular I think there's a possibility that SQL engines may get a bit 'tripped up' by running this on the fly (especially as it'll be running as part of a larger pipeline of code). So there may be mileage in running similar logic on just the array columns earlier on in the sql pipeline, and (temporarily) persisting the results, then having the actual blocking code a bit simpler.

I agree that it's worth experimenting with different ways of implementing this. It's definitely possible that query engines might struggle to execute the current implementation on the fly, but in my investigations with duckdb I have confirmed that it's doing a hash join after it unnests the datasets, which is similar to what it does with regular equality blocking conditions, and probably the most efficient way to do it. The relevant part of the duckdb query plan looks like this:

![image](https://github.com/moj-analytical-services/splink/assets/22582719/1a3df8c0-70b1-44cd-b721-46dcddc302de)

One possible performance issue is that I use a `SELECT DISTINCT` statement, which isn't necessary for more standard blocking rules. This shows up in the query plan as a HASH_GROUP_BY, which may cause memory issues in very large datasets. I'm not sure whether there's a practical way of avoiding the `DISTINCT`.

**Comment by RobinL** [comment]
Thanks that's super useful.

### I was wrong about connected components

@zmbc , good example, should have thought of that!  I had incorrectly assumed that there would be a way of transforming the array column in the input data into a non-array column that can be treated like a normal blocking variable.  But your example and @nerskin’s proposed implementation make it clear to me now that it's not possible.

### SQL performance

Thanks for the additional info @nerskin, the SQL and query plan is really useful.

Re: SQL engines potentially getting tripped up I should have been more specific.  I agree your logic is sound and the joins will be equi-joins (hash joins), and hence should run fast.

Probably it’s easiest to explain the potential issue by example.  It would be possible to compute the full `predict()` pipeline as part of a very large sql statement - essentially a pipeline of many Common Table Expressions.  This would start off with term frequency calcs, then blocking and computation of match probabilities.  

We have found in practice however, that it’s faster to compute the tf tables as separate sql steps and then proceed to blocking because SQL engines don't always like massive SQL pipelines like this.  (Sometimes caching intermediate steps as materialised tables is faster than trying to run the whole thing in one go).

To be clear, The sql code/logic is identical, this is purely about when/where you execute the sql; I'm not suggesting there's anything wrong with your logic.

But overall for simplicity it’s probably best to start with your existing approach, and we can always adjust the sequencing later if we think performance improvements are possible.

### Where to fit this into the Splink codebase

In terms of where this lives in the codebase, we’ve recently done quite a bit of work on the `BlockingRule` class - which lives [here](https://github.com/moj-analytical-services/splink/blob/31d4d0ec9e5c7f27abc25e57968997841a8d004c/splink/blocking.py#L35).  

I’m currently thinking, in terms of the API, it may be best to implement this as part of the `BlockingRule` class - meaning they would still be able to be part of `blocking_rules_to_generate_predictions` in the settings dictionary.  

To see how this may work, it could be similar to the current support for salting in blocking rules.  

Here's an example of salting:

```python

settings = {
    "blocking_rules_to_generate_predictions" = [
        "l.surname = r.surname",
        {"blocking_rule": "l.first_name = r.first_name", "salting_partitions": 10}
    ]
```

Similarly exploding arrays could be something like:

```python
{"blocking_rule": "l.array_col = r.array_col", "explode_arrays": True}
```

In terms of the [`BlockingRule` class](https://github.com/moj-analytical-services/splink/blob/31d4d0ec9e5c7f27abc25e57968997841a8d004c/splink/blocking.py#L35), `explode_arrays` could be introduced as an additional argument to the constructor.

@nerskin, do you think that sounds sensible?

Very happy to support you in moving your existing logic into the right places.  If you'd like to have a go please feel free, or if you start a PR we could move your logic into place.

### A few more things:
- I agree with you that whilst `SELECT DISTINCT` may possibly cause some performance issues, I can't think of better way
- Each blocking rule has a condition that excludes previous blocking rules (see [here](https://github.com/moj-analytical-services/splink/discussions/1417#discussioncomment-6420575)).  I'm wondering whether this may be a bit problematic here (to write the necessary `AND NOT {previous blocking rules}` statement.  A possible fudge may to be allow only one exploding blocking rule, and push it to the bottom of the list of rules.
- With summer hols coming up, apologies if we're slower than we'd like to be in getting this over the line.  It's not because we don't value the contribution and the idea!

**Comment by NickCrews** [comment]
> {"blocking_rule": "l.array_col = r.array_col", "explode_arrays": True}

Could this be made more general? eg `{"blocking_rule": "l.array_vals = r.array_vals", "additional_selections": "unnest(array_col) as array_vals"}`, this selection could get injected into the SQL, similar to @nerskin's example above. It's more flexible, but definitely less clear.

**Comment by NickCrews** [comment]
I'm thinking about the performance implication of this portion of the above SQL:
```
on l.unnested_random_id = r.unnested_random_id and l.gender = r.gender and  l."unique_id" < r."unique_id"
```
Here, the engine has to compute `l.gender = r.gender` many redundant times for a single pair, when we really only need to compute this once.

Can we make it so that
- we do the `l.gender = r.gender and  l."unique_id" < r."unique_id"` on the un-exploded version of the tables and collect the id pairs
- do the `l.unnested_random_id = r.unnested_random_id  and  l."unique_id" < r."unique_id"` on the exploded version and collect the id pairs
- do many union/intersections of these id pairs depending on if the blocking rules are composed with ORs/ANDs.
- I'm not sure where the DISTINCT should occur in the above steps. All of them to reduce the cardinality in each step??
- optionally, execute and store this table of id pairs
- Use this table of id pairs to create `__splink__df_blocked`. This is a hash join so it should be fast, so probably we could save memory/disk by creating this on the fly every time, if the id pairs are the only thing persisted.

**Comment by nerskin** [comment]
Thanks @RobinL, I'll start looking into adding array-intersection blocking rules to the BlockingRule class. The suggestion to let the user specify these blocking rule in a similar way to how salted blocking rules are specified makes sense to me, although in the longer term I think the ideal solution would be to use something like the interface described in https://github.com/moj-analytical-services/splink/issues/1171, which has the advantages of being consistent with how comparisons are specified for the predict step and hiding the fact that salted blocking rules and array-intersection blocking rules work quite differently to the simpler ones. I'm not sure how hard moving to that interface would be though!

> Each blocking rule has a condition that excludes previous blocking rules (see https://github.com/moj-analytical-services/splink/discussions/1417#discussioncomment-6420575), specifically AND NOT (l.first_name = r.first_name) in that example. This may be quite tricky and detrimental to performance (to write the necessary AND NOT {previous blocking rules} statement probably (?) means running the explode logic again. A possible fudge may to be allow only one exploding blocking rule, and push it to the bottom of the list of rules. Or could this be an argument for pre-processing the explode statement, caching the result, and using the resultant table. (similar to how we precompute term frequencies)

The current proof of concept just uses the fudge approach where the array-intersection blocking rules go last and only one of them is allowed, but I'd like to move away from this. I think the simplest thing to do here is to go back to using the tricks that `array_intersection_at_sizes` use to calculate the array overlap. This avoids the need to unnest again, and the performance reasons to use unnesting in the array-intersection blocking rules don't apply here, since the SQL conditions that exclude record pairs from previous blocking rules always result in filtering the set of pairs returned by each blocking rule.

**Comment by nerskin** [comment]
> EDIT: I see @RobinL brought this up in his last comment `...means running the explode logic again...`. I'm more concerned with the impact of the many many redundant comparisons than the explode logic, but perhaps this is what he meant.
> 
> I'm thinking about the performance implication of this portion of the above example SQL:
> 
> ```
> on l.unnested_random_id = r.unnested_random_id and l.gender = r.gender and  l."unique_id" < r."unique_id"
> ```
> 
> Here, the engine has to compute `l.gender = r.gender` many redundant times for a single pair, when we really only need to compute this once.
> 
> Can we make it so that
> 
> * we do the `l.gender = r.gender and  l."unique_id" < r."unique_id"` on the un-exploded version of the tables and collect the id pairs
> * do the `l.unnested_random_id = r.unnested_random_id  and  l."unique_id" < r."unique_id"` on the exploded version and collect the id pairs
> * do many union/intersections of these id pairs depending on if the blocking rules are composed with ORs/ANDs.
> * I'm not sure where the DISTINCT should occur in the above steps. All of them to reduce the cardinality in each step??
> * optionally, execute and store this table of id pairs
> * Use this table of id pairs to create `__splink__df_blocked`. This is a hash join so it should be fast, so probably we could save memory/disk by creating this on the fly every time, if the id pairs are the only thing persisted.

Doing multiple redundant comparisons isn't ideal, but I think that this proposal is likely to make things worse, since collecting the id pairs generated by the non-array blocking variables would often result in the creation of a very large intermediate table (in this example, where it's just `l.gender=r.gender` and `l.unique_id < r.unique_id`, the number of pairs generated for a table with n rows would be something like n^2/4)

**Comment by NickCrews** [comment]
> (in this example, where it's just l.gender=r.gender and l.unique_id < r.unique_id, the number of pairs generated for a table with rows would be something like n^2/4)

Riiiiiiight, thanks. I agree, that is a nonstarter. We need to keep all AND conditions together in the same JOIN step. The rest of what I'm talking about, joining together predicates with ORs, is already what is happening I realize as well.

**Comment by zmbc** [comment]
Is this closed by https://github.com/moj-analytical-services/splink/pull/1692?

**Comment by NickCrews** [comment]
Cross posting here that [I have this implemented in mismo](https://nickcrews.github.io/mismo/reference/block/#mismo.block.block_one), if anyone is trying to implement this themselves. I really like the API I have there, still need to explore more use cases to ensure that my method there will generate fast SQL.

**Comment by RobinL** [comment]
@zmbc yes, good spot, thanks

Closed by #1692
